File: app/src/main/java/com/vikho305/isaho220/outstanding/ui/viewmodel/RegisterViewModel.java
package com.vikho305.isaho220.outstanding.ui.viewmodel;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.vikho305.isaho220.outstanding.data.repositories.UserRepository;
import com.vikho305.isaho220.outstanding.ui.base.BaseViewModel;
import com.vikho305.isaho220.outstanding.util.Resource;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class RegisterViewModel extends BaseViewModel {
    private MutableLiveData<Resource<Boolean>> registerStatus = new MutableLiveData<>();

    private UserRepository userRepository;

    public LiveData<Resource<Boolean>> getRegisterStatus() {
        return registerStatus;
    }

    public RegisterViewModel(Context context) {
        userRepository = new UserRepository(context);
    }

    public void register(String username, String email, String password) {
        registerStatus.postValue(Resource.loading(false));
        addDisposable(
                userRepository.register(username, email, password)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                new Consumer<String>() {
                                    @Override
                                    public void accept(String s) throws Exception {
                                        registerStatus.postValue(Resource.success(true));
                                    }
                                },
                                new Consumer<Throwable>() {
                                    @Override
                                    public void accept(Throwable throwable) throws Exception {
                                        registerStatus.postValue(Resource.error(false));
                                    }
                                }
                        )
        );
    }
}

File: app/src/main/java/com/tavro/parslie/outstanding/ui/viewmodel/RegisterViewModel.kt
package com.tavro.parslie.outstanding.ui.viewmodel

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.tavro.parslie.outstanding.data.model.AuthorizationData
import com.tavro.parslie.outstanding.data.repository.UserRepository
import com.tavro.parslie.outstanding.util.Resource
import com.tavro.parslie.outstanding.util.Status
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers

class RegisterViewModel(context: Context): BaseViewModel() {

    private val userRepository = UserRepository(context)

    private val registerResponse: MutableLiveData<Resource<String>> = MutableLiveData()
    fun getRegisterResponse(): LiveData<Resource<String>> = registerResponse

    fun register(email: String, username: String, password: String) {
        registerResponse.value = Resource(Status.LOADING, null)

        val onSuccess: Consumer<String> = Consumer {
            registerResponse.value = Resource(Status.SUCCESS, it)
        }
        val onError: Consumer<Throwable> = Consumer {
            registerResponse.value = Resource(Status.ERROR, null)
        }

        addDisposable(userRepository.register(email, username, password)
            .subscribeOn(Schedulers.io())               // TODO: look what does
            .observeOn(AndroidSchedulers.mainThread())  // TODO: look what does
            .subscribe(onSuccess, onError))
    }
}